// anichart/lib/core/CanvasHelper.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.canvasHelper = exports.CanvasHelper = void 0;
var canvas_1 = require("canvas");
var CanvasRenderer_1 = require("./CanvasRenderer");
var CanvasHelper = /** @class */ (function () {
    function CanvasHelper() {
        this.renderer = CanvasRenderer_1.canvasRenderer;
        if (typeof window == "undefined") {
            // node
            var nodeCanvas = canvas_1.createCanvas(1920, 1080);
            this.renderer.canvas = nodeCanvas;
            this.renderer.ctx = nodeCanvas.getContext("2d");
        }
        else {
            // browser
            var canvas = document.querySelector("canvas");
            if (!canvas) {
                canvas = document.createElement("canvas");
            }
            this.renderer.canvas = canvas;
            this.renderer.ctx = canvas.getContext("2d");
        }
    }
    CanvasHelper.prototype.isPointInPath = function (area, x, d) {
        var _a;
        if (typeof area == "string") {
            area = new Path2D(area);
        }
        return (_a = this.renderer.ctx) === null || _a === void 0 ? void 0 : _a.isPointInPath(area, x, d);
    };
    CanvasHelper.prototype.measure = function (c) {
        this.renderer.ctx.save();
        if (c.type === "Text") {
            return this.measureText(c);
        }
        this.renderer.ctx.restore();
        return { width: 0 };
    };
    CanvasHelper.prototype.measureText = function (c) {
        var _a;
        this.renderer.prerenderText(c);
        var res = this.renderer.ctx.measureText((_a = c.text) !== null && _a !== void 0 ? _a : "");
        this.renderer.ctx.restore();
        return res;
    };
    return CanvasHelper;
}());
exports.CanvasHelper = CanvasHelper;
exports.canvasHelper = new CanvasHelper();
