// anichart/lib/core/ColorPicker.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.colorPicker = void 0;
var ColorPicker = /** @class */ (function () {
    function ColorPicker() {
        this.colorIndex = 0;
        this.colorMap = new Map();
        this.colorScheme = [
            "#D33",
            "#27C",
            "#FB0",
            "#2C8",
            "#FFF",
            "#0CE",
            "#E8A",
            "#DDA",
            "#C86",
            "#F72",
            "#C8C",
            "#BCA",
            "#F27",
        ];
    }
    ColorPicker.prototype.getColor = function (key) {
        var color = this.colorMap.get(key);
        if (!color) {
            color = this.getNewColor();
            this.colorMap.set(key, color);
        }
        return color;
    };
    ColorPicker.prototype.getNewColor = function () {
        return this.colorScheme[this.colorIndex++ % this.colorScheme.length];
    };
    ColorPicker.prototype.setColor = function (key, color) {
        this.colorMap.set(key, color);
    };
    return ColorPicker;
}());
exports.colorPicker = new ColorPicker();
