// anichart/lib/core/Stage.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.Stage = void 0;
var Ani_1 = require("./ani/Ani");
var CanvasRenderer_1 = require("./CanvasRenderer");
var Component_1 = require("./component/Component");
var FFmpeg_1 = require("./FFmpeg");
var Recourse_1 = require("./Recourse");
var d3_1 = require("d3");
var async_1 = require("async");
var canvas_1 = require("canvas");
// Enable Path2D
require("canvas-5-polyfill");
var Stage = /** @class */ (function () {
    function Stage(canvas) {
        this.compRoot = new Component_1.Component();
        this.options = { sec: 5, fps: 30 };
        this.outputOptions = {
            fileName: "output",
            splitSec: 60,
        };
        this.outputConcurrency = 128;
        this.mode = "output";
        this.cFrame = 0;
        this.renderer = new CanvasRenderer_1.CanvasRenderer();
        this.renderer.stage = this;
        if (typeof window === "undefined") {
            var c = canvas_1.createCanvas(1920, 1080);
            this.renderer.setCanvas(c);
        }
        else {
            if (!canvas) {
                canvas = document.createElement("canvas");
                canvas.width = 1920;
                canvas.height = 1080;
                document.body.appendChild(canvas);
            }
            this.renderer.setCanvas(canvas);
        }
        this.sec = 0;
    }
    Object.defineProperty(Stage.prototype, "frame", {
        get: function () {
            return this.cFrame;
        },
        set: function (val) {
            this.cFrame = val;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Stage.prototype, "sec", {
        get: function () {
            return this.cFrame / this.options.fps;
        },
        set: function (val) {
            this.cFrame = val * this.options.fps;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Stage.prototype, "totalFrames", {
        get: function () {
            return this.options.sec * this.options.fps;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Stage.prototype, "canvas", {
        get: function () {
            return this.renderer.canvas;
        },
        enumerable: false,
        configurable: true
    });
    Stage.prototype.addChild = function (child) {
        this.compRoot.children.push(child);
    };
    Stage.prototype.render = function (sec) {
        if (sec) {
            this.sec = sec;
        }
        this.renderer.clean();
        this.renderer.render(this.compRoot, this.compRoot.offsetSec);
    };
    Stage.prototype.loadRecourse = function () {
        return Recourse_1.recourse.setup();
    };
    Stage.prototype.play = function () {
        var _this = this;
        this.loadRecourse().then(function () {
            _this.doPlay();
        });
    };
    Stage.prototype.doPlay = function () {
        var _this = this;
        this.setup();
        if (this.interval) {
            this.interval.stop();
            this.interval = null;
        }
        else if (typeof window === "undefined") {
            // node
            var f = 0;
            var fs = require("fs");
            while (f < this.totalFrames) {
                this.render(f / this.options.fps);
                var data = this.canvas.toBuffer("image/png");
                var outDir = "out";
                if (!fs.existsSync(outDir)) {
                    fs.mkdirSync(outDir);
                }
                var p = outDir + "/" + this.outputOptions.fileName + "-" + f + ".png";
                fs.writeFileSync(p, data);
                console.log(p);
                f++;
            }
        }
        else if (this.output) {
            FFmpeg_1.loadffmpeg().then(function () {
                var partCount = Math.floor(_this.options.sec / _this.outputOptions.splitSec) + 1;
                var part = 0;
                var parts = [];
                while (part++ < partCount) {
                    parts.push(part);
                }
                async_1.eachSeries(parts, function (p, callback) {
                    var frames = [];
                    var picNameList = [];
                    while (_this.cFrame < _this.totalFrames &&
                        _this.cFrame < p * _this.outputOptions.splitSec * _this.options.fps) {
                        _this.cFrame++;
                        frames.push(_this.cFrame);
                    }
                    async_1.eachLimit(frames, _this.outputConcurrency, function (f, cb) {
                        _this.cFrame = f;
                        _this.render();
                        var no = f - (p - 1) * _this.outputOptions.splitSec * _this.options.fps;
                        picNameList.push("output-" + no + ".png");
                        var imageData = _this.renderer.getImageData();
                        FFmpeg_1.addFrameToFFmpeg(imageData, no).then(function () { return cb(); });
                    }).then(function () {
                        FFmpeg_1.outputMP4(_this.options.fps).then(function () {
                            FFmpeg_1.removePNG(picNameList);
                            callback();
                        });
                    });
                })
                    // tslint:disable-next-line:no-console
                    .then(function () { return console.log("finished!"); });
            });
        }
        else {
            this.interval = d3_1.interval(function (elapsed) {
                if (_this.output || _this.mode === "output") {
                    _this.cFrame++;
                }
                else {
                    _this.cFrame = Math.floor((elapsed / 1000) * _this.options.fps);
                }
                _this.render();
                if (_this.cFrame >= _this.totalFrames) {
                    _this.interval.stop();
                }
            }, (1 / this.options.fps) * 1000);
        }
    };
    Stage.prototype.setup = function () {
        this.setupChildren(this.compRoot);
    };
    Stage.prototype.setupChildren = function (ani) {
        var _this = this;
        ani.setup(this);
        if (ani instanceof Component_1.Component) {
            ani.children.forEach(function (child) {
                if (child instanceof Ani_1.Ani) {
                    _this.setupChildren(child);
                }
            });
        }
    };
    return Stage;
}());
exports.Stage = Stage;
