// anichart/lib/core/chart/BarChart.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BarChart = void 0;
var Component_1 = require("../component/Component");
var Image_1 = require("../component/Image");
var Rect_1 = require("../component/Rect");
var Text_1 = require("../component/Text");
var ColorPicker_1 = require("../ColorPicker");
var CanvasHelper_1 = require("../CanvasHelper");
var BaseChart_1 = require("./BaseChart");
var Recourse_1 = require("../Recourse");
var Constant_1 = require("../Constant");
var d3_1 = require("d3");
var BarChart = /** @class */ (function (_super) {
    __extends(BarChart, _super);
    function BarChart(options) {
        var _this = _super.call(this, options) || this;
        _this.itemCount = 20;
        _this.barPadding = 8;
        _this.barGap = 8;
        _this.swap = 0.25;
        _this.lastValue = new Map();
        _this.dateLabelSize = 60;
        _this.showDateLabel = true;
        _this.barInfoFormat = function (id, data, meta) {
            return _this.labelFormat(id, data, meta);
        };
        if (!options)
            return _this;
        if (options.itemCount)
            _this.itemCount = options.itemCount;
        if (options.barPadding !== undefined)
            _this.barPadding = options.barPadding;
        if (options.barGap !== undefined)
            _this.barGap = options.barGap;
        if (options.barInfoFormat !== undefined)
            _this.barInfoFormat = options.barInfoFormat;
        if (options.dateLabelSize !== undefined)
            _this.dateLabelSize = options.dateLabelSize;
        if (options.showDateLabel !== undefined)
            _this.showDateLabel = options.showDateLabel;
        return _this;
    }
    Object.defineProperty(BarChart.prototype, "sampling", {
        get: function () {
            if (this.stage) {
                return Math.round(this.stage.options.fps * this.swap);
            }
            else {
                return Math.round(30 * this.swap);
            }
        },
        enumerable: false,
        configurable: true
    });
    BarChart.prototype.setup = function (stage) {
        _super.prototype.setup.call(this, stage);
        // 获得曾出现过的Label集合
        this.setShowingIDList();
        this.labelPlaceholder = this.maxLabelWidth;
        this.valuePlaceholder = this.maxValueLabelWidth;
        this.setHistoryIndex();
    };
    BarChart.prototype.setShowingIDList = function () {
        var _this = this;
        var idSet = new Set();
        this.dataGroupByDate.forEach(function (_, date) {
            var dt = _this.secToDate.invert(date);
            __spread(_this.dataScales.entries()).filter(function (d) {
                var data = d[1](dt);
                return data != undefined && !isNaN(data[_this.valueField]);
            })
                .sort(function (a, b) {
                return b[1](dt)[_this.valueField] - a[1](dt)[_this.valueField];
            })
                .slice(0, _this.itemCount)
                .forEach(function (item) {
                var id = item[0];
                idSet.add(id);
            });
        });
        this.IDList = __spread(idSet.values());
    };
    BarChart.prototype.setHistoryIndex = function () {
        var _this = this;
        var dataRange = d3_1.range(this.aniTime[0] - this.swap, this.aniTime[0], this.swap / this.sampling);
        var data = dataRange.map(function (t) {
            return _this.getCurrentData(t).map(function (v) { return v[_this.idField]; });
        });
        this.historyIndex = this.IDList.reduce(function (d, id) {
            var e_1, _a;
            var indexList = [];
            try {
                for (var data_1 = __values(data), data_1_1 = data_1.next(); !data_1_1.done; data_1_1 = data_1.next()) {
                    var dataList = data_1_1.value;
                    var index = dataList.indexOf(id);
                    if (index === -1)
                        index = _this.itemCount;
                    indexList.push(index);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (data_1_1 && !data_1_1.done && (_a = data_1.return)) _a.call(data_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            d.set(id, indexList);
            return d;
        }, new Map());
    };
    Object.defineProperty(BarChart.prototype, "maxValueLabelWidth", {
        get: function () {
            var _this = this;
            var _a;
            var d = __spread(this.data.values());
            var maxWidth = (_a = d3_1.max(d, function (item) {
                var text = new Text_1.Text(_this.getLabelTextOptions(_this.valueFormat(item), "#FFF", _this.barHeight * 0.8));
                var result = CanvasHelper_1.canvasHelper.measure(text);
                return result.width;
            })) !== null && _a !== void 0 ? _a : 0;
            return maxWidth;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BarChart.prototype, "maxLabelWidth", {
        get: function () {
            var _this = this;
            var _a;
            var maxWidth = (_a = d3_1.max(this.IDList, function (id) {
                var text = new Text_1.Text(_this.getLabelTextOptions(_this.labelFormat(id, _this.meta, _this.dataGroupByID), "#FFF", _this.barHeight * 0.8));
                var result = CanvasHelper_1.canvasHelper.measure(text);
                return result.width;
            })) !== null && _a !== void 0 ? _a : 0;
            return maxWidth;
        },
        enumerable: false,
        configurable: true
    });
    BarChart.prototype.getComponent = function (sec) {
        var e_2, _a;
        var _this = this;
        var currentData = this.getCurrentData(sec);
        currentData.forEach(function (d, i) {
            var _a;
            var index = Number.isNaN(d[_this.valueField]) ? _this.itemCount : i;
            (_a = _this.historyIndex.get(d[_this.idField])) === null || _a === void 0 ? void 0 : _a.push(index);
        });
        try {
            for (var _b = __values(this.historyIndex.values()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var history_1 = _c.value;
                var len = history_1.length;
                if (len === this.sampling) {
                    history_1.push(this.itemCount);
                }
                history_1.shift();
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        var indexes = this.IDList.reduce(function (map, id) {
            return map.set(id, d3_1.mean(_this.historyIndex.get(id).map(function (data) { return data; })));
        }, new Map());
        var scaleX;
        if (this.visualRange != "history") {
            var _d = __read(d3_1.extent(currentData, function (d) { return d[_this.valueField]; }), 2), _1 = _d[0], max_1 = _d[1];
            scaleX = d3_1.scaleLinear([0, max_1], [
                0,
                this.shape.width -
                    this.margin.left -
                    this.barPadding -
                    this.labelPlaceholder -
                    this.margin.right -
                    this.valuePlaceholder,
            ]);
        }
        else {
            scaleX = d3_1.scaleLinear([0, d3_1.max(this.data, function (d) { return d[_this.valueField]; })], [
                0,
                this.shape.width -
                    this.margin.left -
                    this.barPadding -
                    this.labelPlaceholder -
                    this.margin.right -
                    this.valuePlaceholder,
            ]);
        }
        var res = new Component_1.Component({
            alpha: this.alphaScale(sec),
            position: this.position,
        });
        currentData.forEach(function (data) {
            var barOptions = _this.getBarOptions(data, scaleX, indexes);
            if (barOptions.alpha > 0) {
                res.children.push(_this.getBarComponent(barOptions));
            }
        });
        if (this.showDateLabel) {
            var dateLabel = new Text_1.Text({
                text: d3_1.timeFormat(this.dateFormat)(this.secToDate(sec)),
                font: Constant_1.font,
                fontSize: this.dateLabelSize,
                fillStyle: "#777",
                textAlign: "right",
                fontWeight: "bolder",
                textBaseline: "bottom",
                position: {
                    x: this.shape.width - this.margin.right,
                    y: this.shape.height - this.margin.bottom,
                },
            });
            res.children.push(dateLabel);
        }
        return res;
    };
    Object.defineProperty(BarChart.prototype, "barHeight", {
        get: function () {
            return ((this.shape.height -
                this.margin.top -
                this.margin.bottom -
                this.barGap * (this.itemCount - 1)) /
                this.itemCount);
        },
        enumerable: false,
        configurable: true
    });
    BarChart.prototype.getBarOptions = function (data, scaleX, indexes) {
        if (!Number.isNaN(data[this.valueField])) {
            this.lastValue.set(data[this.idField], data[this.valueField]);
        }
        data[this.valueField] = this.lastValue.get(data[this.idField]);
        var alpha = d3_1.scaleLinear([this.itemCount - 1, this.itemCount], [1, 0]).clamp(true)(indexes.get(data[this.idField]));
        var color;
        if (typeof this.colorField === "string") {
            color = data[this.idField];
        }
        else {
            color = this.colorField(data[this.idField], this.meta, this.dataGroupByID);
        }
        var image = typeof this.imageField === "string"
            ? data[this.imageField]
            : this.imageField(data[this.idField], this.meta, this.dataGroupByID);
        return {
            id: data[this.idField],
            pos: {
                x: this.margin.left + this.barPadding + this.labelPlaceholder,
                y: this.margin.top +
                    indexes.get(data[this.idField]) * (this.barHeight + this.barGap),
            },
            alpha: alpha,
            data: data,
            image: image,
            value: data[this.valueField],
            shape: { width: scaleX(data[this.valueField]), height: this.barHeight },
            color: ColorPicker_1.colorPicker.getColor(color),
            radius: 4,
        };
    };
    BarChart.prototype.getBarComponent = function (options) {
        var res = new Component_1.Component({
            position: options.pos,
            alpha: options.alpha,
        });
        var bar = new Rect_1.Rect({
            shape: options.shape,
            fillStyle: options.color,
            radius: options.radius,
            clip: true,
        });
        var label = new Text_1.Text(this.getLabelTextOptions(this.labelFormat(options.id, this.meta, this.dataGroupByID), options.color, options.shape.height * 0.8));
        var valueLabel = new Text_1.Text({
            textBaseline: "bottom",
            text: "" + this.valueFormat(options.data),
            textAlign: "left",
            position: {
                x: options.shape.width + this.barPadding,
                y: options.shape.height,
            },
            fontSize: options.shape.height * 0.8,
            font: Constant_1.font,
            fillStyle: options.color,
        });
        var imagePlaceholder = options.image && Recourse_1.recourse.images.get(options.image)
            ? options.shape.height
            : 0;
        var barInfo = new Text_1.Text({
            textAlign: "right",
            textBaseline: "bottom",
            text: this.barInfoFormat(options.id, this.meta),
            position: {
                x: options.shape.width - this.barPadding - imagePlaceholder,
                y: options.shape.height,
            },
            fontSize: options.shape.height * 0.8,
            font: Constant_1.font,
            fontWeight: "bolder",
            fillStyle: "#1e1e1e",
        });
        if (options.image && Recourse_1.recourse.images.get(options.image)) {
            var img = new Image_1.Image({
                src: options.image,
                position: {
                    x: options.shape.width - options.shape.height,
                    y: 0,
                },
                shape: {
                    width: options.shape.height,
                    height: options.shape.height,
                },
            });
            bar.children.push(img);
        }
        bar.children.push(barInfo);
        res.children.push(bar);
        res.children.push(valueLabel);
        res.children.push(label);
        return res;
    };
    BarChart.prototype.getLabelTextOptions = function (text, color, fontSize) {
        if (color === void 0) { color = "#fff"; }
        if (fontSize === void 0) { fontSize = 16; }
        return {
            text: "" + text,
            textAlign: "right",
            textBaseline: "bottom",
            fontSize: fontSize,
            font: Constant_1.font,
            position: { x: 0 - this.barPadding, y: fontSize / 0.8 },
            fillStyle: color,
        };
    };
    return BarChart;
}(BaseChart_1.BaseChart));
exports.BarChart = BarChart;
