// anichart/lib/core/chart/LineChart.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LineChart = void 0;
var CanvasHelper_1 = require("../CanvasHelper");
var ColorPicker_1 = require("../ColorPicker");
var Arc_1 = require("../component/Arc");
var Component_1 = require("../component/Component");
var Path_1 = require("../component/Path");
var Rect_1 = require("../component/Rect");
var Text_1 = require("../component/Text");
var BaseChart_1 = require("./BaseChart");
var Constant_1 = require("../Constant");
var d3_1 = require("d3");
var LineChart = /** @class */ (function (_super) {
    __extends(LineChart, _super);
    function LineChart(options) {
        var _a;
        var _this = _super.call(this, options) || this;
        _this.labelPlaceholder = 0;
        _this.labelSize = 32;
        _this.labelPadding = 4;
        _this.topN = 5;
        if (!options)
            return _this;
        _this.pointerR = (_a = options.pointerR) !== null && _a !== void 0 ? _a : 10;
        return _this;
    }
    LineChart.prototype.setup = function (stage) {
        var _this = this;
        var _a;
        _super.prototype.setup.call(this, stage);
        this.xTickFormat = function (n) {
            return d3_1.timeFormat("%Y-%m-%d")(_this.secToDate(n));
        };
        // Calculate label placeholder
        var textModel = new Text_1.Text({
            fontSize: this.labelSize,
            font: Constant_1.font,
        });
        var labelMaxWidth = (_a = d3_1.max(this.data, function (d) {
            var _a, _b;
            textModel.text = _this.labelFormat(d[_this.idField], _this.meta, d);
            return (_b = (_a = CanvasHelper_1.canvasHelper.measure(textModel)) === null || _a === void 0 ? void 0 : _a.width) !== null && _b !== void 0 ? _b : 0;
        })) !== null && _a !== void 0 ? _a : 0;
        this.labelPlaceholder = labelMaxWidth;
    };
    LineChart.prototype.getComponent = function (sec) {
        var _this = this;
        var res = new Component_1.Component({
            position: this.position,
            alpha: this.alphaScale(sec - this.fadeTime[0] - this.freezeTime[0]),
        });
        if (this.aniTime[0] > sec)
            return null;
        this.scales = this.getScalesBySec(sec);
        var _a = this.getAxis(sec, this.scales), xAxis = _a.xAxis, yAxis = _a.yAxis;
        var lineGen = d3_1.line()
            .defined(function (d) { return !isNaN(d[_this.valueField]); })
            .x(function (d) { return _this.scales.x(_this.secToDate.invert(d[_this.dateField])); })
            .y(function (d) { return _this.scales.y(d[_this.valueField]); });
        var areaGen = d3_1.area()
            .defined(function (d) { return !isNaN(d[_this.valueField]); })
            .x(function (d) { return _this.scales.x(_this.secToDate.invert(d[_this.dateField])); })
            .y0(this.shape.height)
            .y1(function (d) { return _this.scales.y(d[_this.valueField]); });
        var lineArea = new Rect_1.Rect({
            clip: true,
            position: {
                x: this.margin.left + this.yAxisWidth + this.xAxisPadding,
                y: this.margin.top + this.xAxisHeight + this.yAxisPadding,
            },
            shape: {
                width: this.shape.width -
                    this.margin.left -
                    this.margin.right -
                    this.yAxisWidth -
                    this.yAxisPadding -
                    this.labelPlaceholder -
                    this.labelPadding -
                    this.pointerR,
                height: this.shape.height -
                    this.margin.top -
                    this.margin.bottom -
                    this.xAxisHeight -
                    this.xAxisPadding,
            },
            fillStyle: "#0000",
        });
        var points = new Component_1.Component({
            position: {
                x: this.margin.left + this.yAxisWidth + this.yAxisPadding,
                y: this.margin.top + this.xAxisHeight + this.xAxisPadding,
            },
        });
        var labels = new Component_1.Component({
            position: {
                x: this.margin.left + this.yAxisWidth + this.yAxisPadding,
                y: this.margin.top + this.xAxisHeight + this.xAxisPadding,
            },
        });
        var maxX = d3_1.max(this.scales.x.range());
        // 找不到最大值说明啥数据都没有，直接返回
        if (!maxX)
            return res;
        this.dataGroupByID.forEach(function (v, k) {
            var line = new Path_1.Path();
            var color = ColorPicker_1.colorPicker.getColor(k);
            line.strokeStyle = color;
            line.path = lineGen.curve(d3_1.curveMonotoneX)(v);
            line.lineWidth = 3;
            lineArea.children.push(line);
            var areaPath = areaGen.curve(d3_1.curveMonotoneX)(v);
            // 如果画不出Path直接返回
            if (!areaPath)
                return;
            var currentY = _this.findY(areaPath, maxX);
            var point = new Arc_1.Arc({
                fillStyle: color,
                radius: _this.pointerR,
                alpha: currentY !== undefined ? 1 : 0,
                position: { x: maxX, y: currentY },
            });
            var maxValue = _this.scales.y.invert(currentY);
            if (maxValue > _this.historyMax) {
                _this.historyMax = maxValue;
            }
            else if (maxValue < _this.historyMin) {
                _this.historyMin = maxValue;
            }
            points.children.push(point);
            var data = new Map();
            data.set(_this.valueField, maxValue);
            var label = new Text_1.Text({
                text: _this.labelFormat(k, _this.meta, data),
                fontSize: _this.labelSize,
                font: Constant_1.font,
                textAlign: "left",
                textBaseline: "middle",
                position: { x: maxX + _this.labelPadding + _this.pointerR, y: currentY },
                fillStyle: color,
            });
            labels.children.push(label);
        });
        res.children.push(lineArea);
        res.children.push(points);
        res.children.push(xAxis);
        res.children.push(yAxis);
        res.children.push(labels);
        return res;
    };
    LineChart.prototype.getScalesBySec = function (sec) {
        var _this = this;
        var currentData = this.getCurrentData(sec);
        var valueRange = d3_1.extent(currentData, function (d) { return d[_this.valueField]; });
        if (valueRange[0] == undefined) {
            valueRange = [0, 0];
        }
        if (this.historyMax > valueRange[1]) {
            valueRange[1] = this.historyMax;
        }
        if (this.historyMin < valueRange[0]) {
            valueRange[0] = this.historyMin;
        }
        var delta = (valueRange[1] - valueRange[0]) * 0.1;
        valueRange[0] -= delta;
        valueRange[1] += delta;
        var trueSec = sec < this.aniTime[0]
            ? this.aniTime[0]
            : sec > this.aniTime[1]
                ? this.aniTime[1]
                : sec;
        var scales = {
            x: d3_1.scaleLinear([this.aniTime[0], trueSec], [
                0,
                this.shape.width -
                    this.margin.left -
                    this.margin.right -
                    this.yAxisWidth -
                    this.yAxisPadding -
                    this.labelPlaceholder -
                    this.labelPadding -
                    this.pointerR,
            ]),
            y: d3_1.scaleLinear(valueRange, [
                this.shape.height -
                    this.margin.top -
                    this.margin.bottom -
                    this.xAxisHeight,
                this.xAxisPadding,
            ]),
        };
        return scales;
    };
    LineChart.prototype.findY = function (area, x) {
        var l = 0;
        var r = this.shape.height;
        // 9w => 4k
        // 使用中值优化，提升>22倍的性能
        var b = d3_1.bisector(function (d) {
            return CanvasHelper_1.canvasHelper.isPointInPath(area, x, d);
        }).left;
        var yRange = d3_1.range(l, r, 1);
        var index = b(yRange, true);
        return yRange[index];
    };
    return LineChart;
}(BaseChart_1.BaseChart));
exports.LineChart = LineChart;
