// anichart/lib/core/chart/MapChart.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MapChart = void 0;
var d3_1 = require("d3");
var CanvasHelper_1 = require("../CanvasHelper");
var Component_1 = require("../component/Component");
var Path_1 = require("../component/Path");
var Rect_1 = require("../component/Rect");
var Text_1 = require("../component/Text");
var Recourse_1 = require("../Recourse");
var BaseChart_1 = require("./BaseChart");
var MapChart = /** @class */ (function (_super) {
    __extends(MapChart, _super);
    function MapChart(options) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
        var _this = _super.call(this, options) || this;
        _this.labelFormat = function (id) {
            return id;
        };
        if (!options)
            options = {};
        _this.margin = (_a = options === null || options === void 0 ? void 0 : options.margin) !== null && _a !== void 0 ? _a : {
            top: 20,
            left: 20,
            right: 20,
            bottom: 20,
        };
        _this.visualMap = (_b = options.visualMap) !== null && _b !== void 0 ? _b : d3_1.interpolateInferno;
        _this.getMapId = (_c = options.getMapId) !== null && _c !== void 0 ? _c : (function (id) { return id; });
        _this.mapIdField = (_d = options.mapIdField) !== null && _d !== void 0 ? _d : "alpha3Code";
        _this.strokeStyle = (_e = options.strokeStyle) !== null && _e !== void 0 ? _e : "#FFF";
        _this.defaultFill = (_f = options.defaultFill) !== null && _f !== void 0 ? _f : "#FFF1";
        _this.projectionType = (_g = options.projectionType) !== null && _g !== void 0 ? _g : "natural";
        _this.useShadow = (_h = options.useShadow) !== null && _h !== void 0 ? _h : false;
        _this.pathShadowColor = options.pathShadowColor;
        _this.pathShadowBlur = (_j = options.pathShadowBlur) !== null && _j !== void 0 ? _j : 100;
        _this.showGraticule = (_k = options.showGraticule) !== null && _k !== void 0 ? _k : false;
        _this.showLabel = (_l = options.showLabel) !== null && _l !== void 0 ? _l : false;
        _this.noDataLabel = (_m = options.noDataLabel) !== null && _m !== void 0 ? _m : undefined;
        _this.labelPadding = (_o = options.labelPadding) !== null && _o !== void 0 ? _o : 8;
        _this.labelSize = (_p = options.labelSize) !== null && _p !== void 0 ? _p : 12;
        if (options.labelFormat)
            _this.labelFormat = options.labelFormat;
        _this.labelAlphaScale = (_q = options.labelAlphaScale) !== null && _q !== void 0 ? _q : d3_1.scaleLinear([400, 560], [0, 1]).clamp(true);
        return _this;
    }
    MapChart.prototype.setup = function (stage) {
        _super.prototype.setup.call(this, stage);
        if (stage) {
            var map = Recourse_1.recourse.data.get("map");
            var projection = void 0;
            switch (this.projectionType) {
                case "orthographic":
                    projection = d3_1.geoOrthographic();
                    break;
                case "natural":
                    projection = d3_1.geoNaturalEarth1();
                    break;
                case "mercator":
                    projection = d3_1.geoMercator();
                    break;
                case "equirectangular":
                    projection = d3_1.geoEquirectangular();
                default:
                    projection = d3_1.geoNaturalEarth1();
            }
            projection.fitExtent([
                [this.margin.left, this.margin.top],
                [
                    stage.canvas.width - this.margin.right,
                    stage.canvas.height - this.margin.bottom,
                ],
            ], map);
            this.projection = projection;
            this.map = map;
            this.init(projection, map);
        }
    };
    MapChart.prototype.init = function (projection, map) {
        this.initGeoPath(projection, map);
        this.initComps();
    };
    MapChart.prototype.initGeoPath = function (projection, map) {
        var geoGener = d3_1.geoPath(projection);
        this.geoGener = geoGener;
        this.pathMap = new Map();
        this.initPathMap(map, geoGener);
    };
    MapChart.prototype.initPathMap = function (map, geoGener) {
        var e_1, _a;
        this.labelComponentMap = new Map();
        try {
            for (var _b = __values(map.features), _c = _b.next(); !_c.done; _c = _b.next()) {
                var feature = _c.value;
                var mapId = feature.properties[this.mapIdField];
                var path = geoGener(feature);
                this.pathMap.set(mapId, path);
                var txt = new Text_1.Text({
                    position: { x: 4, y: 6 },
                    text: this.labelFormat(feature.properties[this.mapIdField]),
                    textAlign: "left",
                    textBaseline: "top",
                    fillStyle: this.strokeStyle,
                    fontSize: this.labelSize,
                });
                var width = CanvasHelper_1.canvasHelper.measure(txt).width;
                var label = new Rect_1.Rect({
                    position: { x: 0, y: 0 },
                    fillStyle: "#2225",
                    strokeStyle: this.strokeStyle,
                    shape: {
                        width: width + this.labelPadding,
                        height: this.labelSize + this.labelPadding,
                    },
                });
                label.addChild(txt);
                this.labelComponentMap.set(mapId, label);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    MapChart.prototype.initComps = function () {
        var e_2, _a;
        var _this = this;
        this.wrapper = new Component_1.Component();
        this.pathComponentMap = new Map();
        this.pathMap.forEach(function (p, mapId) {
            var path = new Path_1.Path({
                path: p,
                fillStyle: _this.defaultFill,
                strokeStyle: _this.strokeStyle,
            });
            _this.wrapper.children.push(path);
            _this.pathComponentMap.set(mapId, path);
        });
        if (this.showLabel) {
            try {
                for (var _b = __values(this.labelComponentMap.values()), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var labelComp = _c.value;
                    this.wrapper.children.push(labelComp);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        if (this.showGraticule) {
            var stroke = d3_1.color(this.strokeStyle);
            if (stroke) {
                stroke.opacity = 0.25;
            }
            this.graticulePathComp = new Path_1.Path({
                path: this.graticulePath,
                strokeStyle: stroke === null || stroke === void 0 ? void 0 : stroke.toString(),
                fillStyle: "#0000",
            });
            this.wrapper.children.push(this.graticulePathComp);
        }
    };
    MapChart.prototype.getComponent = function (sec) {
        this.updateScale(sec);
        this.updateProject(sec);
        this.updatePath(sec);
        return this.wrapper;
    };
    MapChart.prototype.updateScale = function (sec) {
        var _a;
        var _this = this;
        _a = __read(d3_1.extent(this.getCurrentData(sec), function (d) { return d[_this.valueField]; }), 2), this.currentMin = _a[0], this.currentMax = _a[1];
        if (this.currentMax > this.historyMax) {
            this.historyMax = this.currentMax;
        }
        if (this.historyMin > this.currentMin) {
            this.historyMin = this.currentMax;
        }
        if (!this.visualRange || typeof this.visualRange === "string") {
            switch (this.visualRange) {
                case "total":
                    this.scale = d3_1.scaleLinear([this.totallyMin, this.totallyMax], [0, 1]).clamp(true);
                    break;
                case "history":
                    this.scale = d3_1.scaleLinear([this.historyMin, this.historyMax], [0, 1]).clamp(true);
                default:
                    this.scale = d3_1.scaleLinear([this.currentMin, this.currentMax], [0, 1]).clamp(true);
                    break;
            }
        }
        else {
            this.scale = d3_1.scaleLinear(this.visualRange, [0, 1]).clamp(true);
        }
    };
    MapChart.prototype.updatePath = function (sec) {
        var e_3, _a, e_4, _b;
        var _c, _d, _e;
        if (this.showGraticule) {
            var graticulePath = this.geoGener(d3_1.geoGraticule10());
            if (graticulePath) {
                this.graticulePathComp.path = graticulePath;
            }
        }
        try {
            for (var _f = __values(this.map.features), _g = _f.next(); !_g.done; _g = _f.next()) {
                var feature = _g.value;
                var mapId = feature.properties[this.mapIdField];
                var path = this.geoGener(feature);
                var comp = this.pathComponentMap.get(mapId);
                if (comp && path) {
                    comp.path = path;
                }
                var label = this.labelComponentMap.get(mapId);
                if (label) {
                    var center = this.geoGener.centroid(feature);
                    var area = this.geoGener.area(feature);
                    label.alpha = path ? this.labelAlphaScale(area) : 0;
                    label.position.x = center[0];
                    label.position.y = center[1];
                }
                if (label) {
                    label.children[0].text =
                        ((_d = (_c = this.dataScales) === null || _c === void 0 ? void 0 : _c.get(mapId)) === null || _d === void 0 ? void 0 : _d(sec)[this.valueField]) ||
                            !this.noDataLabel
                            ? this.labelFormat(mapId)
                            : this.noDataLabel;
                    var width = CanvasHelper_1.canvasHelper.measure(label.children[0]).width;
                    label.shape.width = width + this.labelPadding;
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_g && !_g.done && (_a = _f.return)) _a.call(_f);
            }
            finally { if (e_3) throw e_3.error; }
        }
        try {
            for (var _h = __values(this.dataScales), _j = _h.next(); !_j.done; _j = _h.next()) {
                var _k = __read(_j.value, 2), id = _k[0], data = _k[1];
                var mapId = this.getMapId(id);
                var currentValue = data(sec)[this.valueField];
                var rate = this.scale(currentValue);
                var color_1 = this.visualMap(rate);
                var comp = this.pathComponentMap.get(mapId);
                if (comp) {
                    comp.fillStyle = currentValue ? color_1 : this.defaultFill;
                    if (this.useShadow) {
                        comp.shadow = {
                            enable: true,
                            blur: this.pathShadowBlur * rate + 1,
                            color: (_e = this.pathShadowColor) !== null && _e !== void 0 ? _e : color_1,
                        };
                    }
                }
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_j && !_j.done && (_b = _h.return)) _b.call(_h);
            }
            finally { if (e_4) throw e_4.error; }
        }
    };
    MapChart.prototype.updateProject = function (sec) {
        this.projection.rotate([sec * 20, 0]);
    };
    return MapChart;
}(BaseChart_1.BaseChart));
exports.MapChart = MapChart;
