// anichart/lib/core/chart/PieChart.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PieChart = void 0;
var BaseChart_1 = require("./BaseChart");
var Path_1 = require("../component/Path");
var ColorPicker_1 = require("../ColorPicker");
var Text_1 = require("../component/Text");
var Constant_1 = require("../Constant");
var d3_1 = require("d3");
var PieChart = /** @class */ (function (_super) {
    __extends(PieChart, _super);
    function PieChart(options) {
        var _this = _super.call(this, options) || this;
        _this.minRadio = 0;
        _this.radius = [0, 120];
        _this.cornerRadius = 4;
        _this.padAngle = 5;
        _this.keyDurationSec = 0.25;
        _this.labelTextStyle = {
            font: Constant_1.font,
            lineWidth: 6,
            fontSize: 24,
            fontWeight: "bolder",
            strokeStyle: "#1e1e1e",
        };
        if (options) {
            Object.assign(_this, options);
        }
        return _this;
    }
    PieChart.prototype.getComponent = function (sec) {
        var e_1, _a;
        var res = _super.prototype.getComponent.call(this, sec);
        if (this.showDateLabel) {
            var date = this.secToDate(sec);
            var dateLabel = new Text_1.Text({
                text: d3_1.timeFormat(this.dateFormat)(date),
                fillStyle: "#FFF",
                fontSize: 30,
                position: { x: 0, y: 0 },
            });
            res === null || res === void 0 ? void 0 : res.children.push(dateLabel);
        }
        var remained = sec % this.keyDurationSec;
        var start = sec - remained;
        var end = start + this.keyDurationSec;
        var comp0 = this.getPieData(start);
        var comp1 = this.getPieData(end);
        var pieData = d3_1.scaleLinear([start, end], [comp0, comp1])(remained + start);
        var arcGen = d3_1.arc();
        try {
            for (var pieData_1 = __values(pieData), pieData_1_1 = pieData_1.next(); !pieData_1_1.done; pieData_1_1 = pieData_1.next()) {
                var d = pieData_1_1.value;
                var path = arcGen
                    .endAngle(d.endAngle)
                    .padAngle(d.padAngle)
                    .startAngle(d.startAngle)
                    .innerRadius(this.radius[0])
                    .outerRadius(this.radius[1])
                    .cornerRadius(this.cornerRadius)
                    .padAngle(d.padAngle)({
                    innerRadius: 0,
                    outerRadius: 0,
                    startAngle: 0,
                    endAngle: 0,
                });
                var centroid = arcGen.centroid(pieData);
                var label = new Text_1.Text({
                    text: d.data[this.idField],
                    fontSize: this.labelTextStyle.fontSize,
                    lineWidth: this.labelTextStyle.lineWidth,
                    font: this.labelTextStyle.font,
                    textAlign: "center",
                    textBaseline: "middle",
                    fillStyle: ColorPicker_1.colorPicker.getColor(d.data[this.idField]),
                    fontWeight: this.labelTextStyle.fontWeight,
                    strokeStyle: this.labelTextStyle.strokeStyle,
                    position: { x: centroid[0], y: centroid[1] },
                });
                var comp = new Path_1.Path({
                    fillStyle: ColorPicker_1.colorPicker.getColor(d.data[this.idField]),
                    strokeStyle: "#0000",
                    path: path,
                });
                res === null || res === void 0 ? void 0 : res.children.push(comp);
                res === null || res === void 0 ? void 0 : res.children.push(label);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (pieData_1_1 && !pieData_1_1.done && (_a = pieData_1.return)) _a.call(pieData_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return res;
    };
    PieChart.prototype.getPieData = function (sec) {
        var _this = this;
        var currentData = __spread(this.dataScales.values()).map(function (scale) {
            return scale(sec);
        });
        var minRadio = this.minRadio / 360;
        var sumValue = d3_1.sum(currentData, function (d) { return d[_this.valueField]; });
        var minValue = sumValue * minRadio;
        var pieGen = d3_1.pie()
            .padAngle((Math.PI / 180) * this.padAngle)
            .value(function (d) { return d3_1.max([d[_this.valueField], minValue]); });
        currentData.sort(function (a, b) {
            if (Number.isNaN(b[_this.valueField])) {
                return -1;
            }
            else if (Number.isNaN(a[_this.valueField])) {
                return 1;
            }
            else {
                return b[_this.idField] - a[_this.idField];
            }
        });
        var pieData = pieGen(currentData);
        return pieData;
    };
    return PieChart;
}(BaseChart_1.BaseChart));
exports.PieChart = PieChart;
