// anichart/lib/core/ani/AniCreator.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFadeAni = exports.createAni = exports.customAni = exports.easeInterpolate = void 0;
var Ani_1 = require("./Ani");
var Fade_1 = require("../wrapper/Fade");
var d3_1 = require("d3");
function easeInterpolate(e) {
    return function (a, b) {
        var i = d3_1.interpolate(a, b);
        return function (t) {
            return i(e(t));
        };
    };
}
exports.easeInterpolate = easeInterpolate;
var CustomAniNeedFrame = /** @class */ (function () {
    function CustomAniNeedFrame(startSec) {
        if (startSec === void 0) { startSec = 0; }
        this.keyFrames = [];
        this.keyTimes = [];
        this.eases = [];
        this.keyTimes.push(startSec);
    }
    CustomAniNeedFrame.prototype.keyFrame = function (c) {
        this.keyFrames.push(c);
        return new CustomAni(this, this.keyTimes, this.keyFrames, this.eases);
    };
    return CustomAniNeedFrame;
}());
// tslint:disable-next-line:max-classes-per-file
var CustomAni = /** @class */ (function (_super) {
    __extends(CustomAni, _super);
    function CustomAni(aniSeries, keyTimes, keyFrames, eases) {
        var _this = _super.call(this) || this;
        _this.aniSeries = aniSeries;
        _this.keyTimes = keyTimes;
        _this.keyFrames = keyFrames;
        _this.eases = eases;
        return _this;
    }
    CustomAni.prototype.getComponent = function (sec) {
        // [0, 3, 6]
        var rIdx = d3_1.bisectLeft(this.keyTimes, sec);
        if (rIdx >= this.keyFrames.length) {
            rIdx = this.keyFrames.length - 1;
        }
        var lIdx = rIdx - 1;
        if (lIdx < 0) {
            return null;
        }
        var eIdx = lIdx >= this.eases.length ? this.eases.length - 1 : lIdx;
        var scale = d3_1.scaleLinear([this.keyTimes[lIdx], this.keyTimes[rIdx]], [this.keyFrames[lIdx], this.keyFrames[rIdx]])
            .interpolate(easeInterpolate(this.eases[eIdx]))
            .clamp(true);
        return scale(sec);
    };
    CustomAni.prototype.duration = function (duration, ease) {
        if (ease === void 0) { ease = d3_1.easeLinear; }
        this.keyTimes.push(this.keyTimes[this.keyTimes.length - 1] + duration);
        this.eases.push(ease);
        return this.aniSeries;
    };
    return CustomAni;
}(Ani_1.Ani));
function customAni(startSec) {
    return new CustomAniNeedFrame(startSec ? startSec : 0);
}
exports.customAni = customAni;
function createAni(keyFrames, keyTimes, ease) {
    if (keyTimes === void 0) { keyTimes = [0, 1]; }
    if (ease === void 0) { ease = d3_1.easeLinear; }
    var scale = d3_1.scaleLinear(keyTimes, keyFrames)
        .interpolate(easeInterpolate(ease))
        .clamp(true);
    var ani = new Ani_1.Ani();
    ani.getComponent = function (i) {
        return scale(i);
    };
    return ani;
}
exports.createAni = createAni;
function getFadeAni(obj, options) {
    var _a, _b, _c;
    var startSec = (_a = options === null || options === void 0 ? void 0 : options.startSec) !== null && _a !== void 0 ? _a : 0;
    var durationSec = (_b = options === null || options === void 0 ? void 0 : options.durationSec) !== null && _b !== void 0 ? _b : 3;
    var fadeSec = (_c = options === null || options === void 0 ? void 0 : options.fadeSec) !== null && _c !== void 0 ? _c : 1;
    var alphaScale = d3_1.scaleLinear([
        startSec,
        startSec + fadeSec,
        startSec + durationSec - fadeSec,
        startSec + durationSec,
    ], [0, 1, 1, 0]).clamp(true);
    return Fade_1.getFadeWrapped(obj, alphaScale);
}
exports.getFadeAni = getFadeAni;
