// anichart/lib/core/ani/GridAni.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GridAni = void 0;
var Component_1 = require("../component/Component");
var Ani_1 = require("./Ani");
var GridAni = /** @class */ (function (_super) {
    __extends(GridAni, _super);
    function GridAni(options) {
        var _a, _b, _c, _d, _e;
        var _this = _super.call(this) || this;
        if (options) {
            _this.col = (_a = options.col) !== null && _a !== void 0 ? _a : 3;
            _this.row = (_b = options.row) !== null && _b !== void 0 ? _b : 3;
            _this.position = (_c = options.position) !== null && _c !== void 0 ? _c : { x: 0, y: 0 };
            _this.items = (_d = options.items) !== null && _d !== void 0 ? _d : [];
            _this.shape = (_e = options.shape) !== null && _e !== void 0 ? _e : { width: 0, height: 0 };
        }
        return _this;
    }
    GridAni.prototype.setup = function (stage) {
        var _this = this;
        _super.prototype.setup.call(this, stage);
        if (!this.shape) {
            this.shape = {
                width: stage.canvas.width,
                height: stage.canvas.height,
            };
        }
        var height = this.shape.height / this.row;
        var width = this.shape.width / this.col;
        this.wrapper = new Component_1.Component({
            position: this.position,
        });
        this.items.forEach(function (item, index) {
            var col = index % _this.col;
            var row = Math.floor(index / _this.col);
            item.position = { x: width * col + width / 2, y: height * row };
            _this.wrapper.addChild(item);
        });
    };
    GridAni.prototype.getComponent = function (sec) {
        return this.wrapper;
    };
    return GridAni;
}(Ani_1.Ani));
exports.GridAni = GridAni;
