// anichart/lib/core/ani/Progress.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Progress = void 0;
var Component_1 = require("../component/Component");
var Text_1 = require("../component/Text");
var Rect_1 = require("../component/Rect");
var Ani_1 = require("./Ani");
var AniCreator_1 = require("./AniCreator");
var Constant_1 = require("../Constant");
var d3_1 = require("d3");
var Progress = /** @class */ (function (_super) {
    __extends(Progress, _super);
    function Progress(options) {
        var _a, _b, _c, _d;
        var _this = _super.call(this) || this;
        _this.radius = 6;
        _this.padding = 3;
        _this.lineWidth = 2;
        _this.center = { x: 0, y: 0 };
        if (options) {
            _this.position = (_a = options.position) !== null && _a !== void 0 ? _a : { x: 0, y: 0 };
            _this.aniTime = (_b = options.aniTime) !== null && _b !== void 0 ? _b : [0, 3];
            _this.shape = (_c = options.shape) !== null && _c !== void 0 ? _c : { width: 400, height: 18 };
            _this.color = (_d = options.color) !== null && _d !== void 0 ? _d : "#FFF";
        }
        var border0 = new Rect_1.Rect({
            shape: {
                width: _this.shape.width,
                height: _this.shape.height,
            },
            center: { x: _this.shape.width / 2, y: _this.shape.height / 2 },
            alpha: 1,
            radius: _this.radius,
            fillStyle: "#0006",
            strokeStyle: _this.color,
            lineWidth: _this.lineWidth,
        });
        var border1 = new Rect_1.Rect({
            shape: {
                width: _this.shape.width * 1.75,
                height: _this.shape.height,
            },
            center: {
                x: (_this.shape.width * 1.75) / 2,
                y: _this.shape.height / 2,
            },
            alpha: 1,
            radius: _this.radius,
            fillStyle: "#0006",
            strokeStyle: _this.color,
            lineWidth: _this.lineWidth,
        });
        var bar0 = new Rect_1.Rect({
            position: { x: _this.padding, y: _this.padding },
            center: { x: _this.shape.width / 2, y: _this.shape.height / 2 },
            shape: { width: 0, height: _this.shape.height - _this.padding * 2 },
            radius: _this.radius,
            fillStyle: _this.color,
        });
        var bar1 = new Rect_1.Rect({
            position: { x: _this.padding, y: _this.padding },
            center: { x: (_this.shape.width / 2) * 1.75, y: _this.shape.height / 2 },
            shape: {
                width: (_this.shape.width - _this.padding * 2) * 1.75,
                height: _this.shape.height - _this.padding * 2,
            },
            radius: _this.radius,
            fillStyle: _this.color,
        });
        var start = new Component_1.Component({
            position: _this.position,
            scale: { x: 2, y: 2 },
        });
        start.children.push(border0);
        start.children.push(bar0);
        var end = new Component_1.Component({
            position: _this.position,
            scale: { x: 1, y: 1 },
        });
        end.children.push(border1);
        end.children.push(bar1);
        var borderFinished = new Rect_1.Rect({
            shape: {
                width: _this.shape.width * 2,
                height: _this.shape.height * 2,
            },
            center: {
                x: (_this.shape.width * 2) / 2,
                y: (_this.shape.height * 2) / 2,
            },
            alpha: 0,
            radius: _this.radius,
            fillStyle: "#0006",
            strokeStyle: "#27C",
            lineWidth: _this.lineWidth,
        });
        var bar2 = new Rect_1.Rect({
            center: { x: _this.shape.width / 2, y: _this.shape.height / 2 },
            position: { x: _this.padding, y: _this.padding },
            shape: {
                width: _this.shape.width - _this.padding * 2,
                height: _this.shape.height - _this.padding * 2,
            },
            alpha: 0.0,
            radius: _this.radius,
            fillStyle: "#27C",
        });
        var final = new Component_1.Component({
            position: _this.position,
            alpha: 1,
            scale: { x: 1.2, y: 1.2 },
        });
        final.children.push(borderFinished);
        final.children.push(bar2);
        var objCopy = Object.assign({}, final);
        objCopy.alpha = 0;
        _this.ani = AniCreator_1.customAni(_this.aniTime[0])
            .keyFrame(start)
            .duration(_this.aniTime[1], d3_1.easePolyOut.exponent(5))
            .keyFrame(end)
            .duration(0.25, d3_1.easeExpOut)
            .keyFrame(final)
            .duration(0.5)
            .keyFrame(final)
            .duration(0.2)
            .keyFrame(objCopy);
        return _this;
    }
    Progress.prototype.getComponent = function (sec) {
        var val = d3_1.scaleLinear(this.aniTime, [0, 100])
            .clamp(true)
            .interpolate(AniCreator_1.easeInterpolate(d3_1.easePolyOut.exponent(5)))(sec);
        var label = d3_1.format("d")(val);
        var res = this.ani.getComponent(sec);
        var textLabel = new Text_1.Text({
            text: val === 100 ? "" : "Loading " + label + " %",
            font: Constant_1.font,
            fontSize: 24,
            textAlign: "center",
            textBaseline: "top",
            position: { x: 0, y: this.shape.height },
            fillStyle: "#fff",
        });
        res === null || res === void 0 ? void 0 : res.children.push(textLabel);
        return res;
    };
    return Progress;
}(Ani_1.Ani));
exports.Progress = Progress;
