// anichart/lib/core/custom/TextWithBackGround.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTextWithBackground = void 0;
var CanvasHelper_1 = require("../CanvasHelper");
var Component_1 = require("../component/Component");
var Rect_1 = require("../component/Rect");
var Text_1 = require("../component/Text");
var Constant_1 = require("../Constant");
function getTextWithBackground(_a) {
    var _b;
    var _c = _a.txt, txt = _c === void 0 ? "在此处输入文字！" : _c, _d = _a.position, position = _d === void 0 ? { x: 100, y: 100 } : _d, _e = _a.fontSize, fontSize = _e === void 0 ? 24 : _e, _f = _a.fontWeight, fontWeight = _f === void 0 ? "bold" : _f, _g = _a.textAlign, textAlign = _g === void 0 ? "right" : _g, _h = _a.textBaseline, textBaseline = _h === void 0 ? "top" : _h, _j = _a.foregroundStyle, foregroundStyle = _j === void 0 ? "#1e1e1e" : _j, _k = _a.backgroundStyle, backgroundStyle = _k === void 0 ? "#fff" : _k, _l = _a.padding, padding = _l === void 0 ? { top: 5, bottom: 0, left: 5, right: 5 } : _l;
    var res = new Component_1.Component({
        position: position,
    });
    var text = new Text_1.Text({
        text: txt,
        position: { x: padding.left, y: padding.top },
        fillStyle: foregroundStyle,
        font: Constant_1.font,
        fontSize: fontSize,
        textBaseline: textBaseline,
        fontWeight: fontWeight,
        textAlign: textAlign,
    });
    var width = (_b = CanvasHelper_1.canvasHelper.measure(text)) === null || _b === void 0 ? void 0 : _b.width;
    var rect = new Rect_1.Rect({
        clip: true,
        shape: {
            width: width + text.position.x + padding.left + padding.right,
            height: fontSize + padding.top + padding.bottom,
        },
        fillStyle: backgroundStyle,
    });
    if (textAlign === "right") {
        text.position.x += width;
    }
    res.center = { x: rect.shape.width / 2, y: rect.shape.height / 2 };
    rect.children.push(text);
    res.children.push(rect);
    return res;
}
exports.getTextWithBackground = getTextWithBackground;
