// anichart/lib/core/component/Component.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.Component = void 0;
var Component = /** @class */ (function () {
    function Component(options) {
        this.type = "Component";
        this.offsetSec = 0;
        this.shadow = { enable: false };
        this.center = { x: 0, y: 0 };
        this.offset = { x: 0, y: 0 };
        this.children = [];
        if (options) {
            if (options.offsetSec)
                this.offsetSec = options.offsetSec;
            if (options.center)
                this.center = options.center;
            if (options.shadow)
                this.shadow = options.shadow;
            if (options.position)
                this.position = options.position;
            if (options.alpha !== undefined)
                this.alpha = options.alpha;
            if (options.offset)
                this.offset = options.offset;
            if (options.children)
                this.children = options.children;
            if (options.scale)
                this.scale = options.scale;
            if (options.filter)
                this.filter = options.filter;
            if (options.fillStyle)
                this.fillStyle = options.fillStyle;
            if (options.strokeStyle)
                this.strokeStyle = options.strokeStyle;
            if (options.lineWidth)
                this.lineWidth = options.lineWidth;
        }
    }
    Component.prototype.setup = function (stage) {
        this.children.forEach(function (child) {
            child.setup(stage);
        });
    };
    Component.prototype.addChild = function (comp) {
        this.children.push(comp);
        comp.parent = this;
    };
    return Component;
}());
exports.Component = Component;
