// anichart/lib/core/component/Grid.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Grid = void 0;
var Component_1 = require("./Component");
var Grid = /** @class */ (function (_super) {
    __extends(Grid, _super);
    function Grid(options) {
        var _a, _b, _c;
        var _this = _super.call(this, options) || this;
        _this.col = (_a = options === null || options === void 0 ? void 0 : options.col) !== null && _a !== void 0 ? _a : 3;
        _this.row = (_b = options === null || options === void 0 ? void 0 : options.row) !== null && _b !== void 0 ? _b : undefined;
        _this.shape = (_c = options === null || options === void 0 ? void 0 : options.shape) !== null && _c !== void 0 ? _c : undefined;
        return _this;
    }
    Grid.prototype.setup = function (stage) {
        var _this = this;
        var _a;
        _super.prototype.setup.call(this, stage);
        this.row = (_a = this.row) !== null && _a !== void 0 ? _a : Math.floor(this.children.length / this.col) + 1;
        if (!this.shape) {
            this.shape = {
                width: stage.canvas.width,
                height: stage.canvas.height,
            };
        }
        var height = this.shape.height / this.row;
        var width = this.shape.width / this.col;
        this.children.forEach(function (item, index) {
            var col = index % _this.col;
            var row = Math.floor(index / _this.col);
            if (item instanceof Component_1.Component) {
                item.position = { x: width * col + width / 2, y: height * row };
            }
            else {
                var getComp_1 = item.getComponent;
                item.getComponent = function (sec) {
                    var comp = getComp_1(sec);
                    if (comp) {
                        comp.position = { x: width * col + width / 2, y: height * row };
                    }
                    return comp;
                };
            }
        });
    };
    return Grid;
}(Component_1.Component));
exports.Grid = Grid;
